### ISSUE TITLE
A bug of header adding and getting in OpenFeign RequestInterceptor

### ISSUE THREAD

**Issue by HaojunRen** [OpenFeign/feign#1987]
Refer to the following issue at Spring Cloud community

https://github.com/spring-cloud/spring-cloud-openfeign/issues/760

**Comment by radio-rogal** [comment]
Hi!

Please try to put keys to doublequoters, it works:
```java
  @Test
  public void headerWithJSONValue() {
    RequestTemplate template = new RequestTemplate()
        .header("key", "{\"x\":1.0,\"y\":1.1}");

    assertThat(template.headers()).hasSize(1);
    assertThat(template.headers().keySet()).containsExactly("key");
    assertThat(template.headers().get("key")).containsExactly("{\"x\":1.0,\"y\":1.1}");
  }
```

**Comment by radio-rogal** [comment]
This is JSON's rule: string are wrapped double quotes, both in keys and values.

And as far as I guess this rule is used to avoid processing of header value, see [Request Headers Expansion](https://github.com/OpenFeign/feign#request-body-expansion) that use curly brackets like JSON does.

**Comment by kdavisk6** [comment]
> It does not make sense, that means i should add double-quoters to every header value with json format one by one?

As @radio-rogal has stated, Feign supports expression based headers.  This means you can use URI templates in your header values and have them resolved at execution time.  The incompatibility you are seeing is related to the use of JSON as a header value, which conflicts with the URI Template specification where dynamic expressions are wrapped in braces `{}`.  

If you chose to use Feign and require JSON based header values, you will need to quote them as discussed.

**Comment by yangyang122** [comment]
![1699345955608](https://github.com/OpenFeign/feign/assets/12775735/59808171-7f3a-485f-bddb-330a2298f123)

Can you add a switch here to enable parsing JSON data？

**Comment by yangyang122** [comment]
I try : 
 ` String data="{\"x\":1.0,\"y\":2.0}";
        RequestTemplate rq = requestTemplate.header("v", data);
        Map<String, Collection<String>> headers = rq.headers();
        headers.forEach((key, values) -> {
            System.out.println(key+"--"+values);
        });`
![Uploading 图片.png…]()  
but it's not the expected value.

**Comment by hyfsy** [comment]
> Hi!
> 
> Please try to put keys to doublequoters, it works:
> 
> ```java
>   @Test
>   public void headerWithJSONValue() {
>     RequestTemplate template = new RequestTemplate()
>         .header("key", "{\"x\":1.0,\"y\":1.1}");
> 
>     assertThat(template.headers()).hasSize(1);
>     assertThat(template.headers().keySet()).containsExactly("key");
>     assertThat(template.headers().get("key")).containsExactly("{\"x\":1.0,\"y\":1.1}");
>   }
> ```

It doesn't work in version:
spring-cloud-starter-openfeign: 3.1.4
feign-core: 11.8

**Comment by kdavisk6** [comment]
Fixed in #2471
